### PR TITLE
[D2M] add lowering for comparison ops

### PR DESCRIPTION
**Pull request by dlokeTT**
### Ticket
#3013
#3014 
#3015

### Problem description
Need to support comparison ops: eq, ge, gt for llama 

### What's changed
Added lowering for all comparison ops, (eq, ne, ge, gt, le, lt) for d2m path

### Checklist
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td [hunk 1]
```diff
@@ -279,6 +279,17 @@ def TTIR_TileLogicalNotOp : TTIR_GenericRegionComputeUnaryDstOp<"tile_logical_no
     let results = (outs TTCore_Tile:$result);
 }
 
+def TTIR_TileEqOp : TTIR_GenericRegionComputeOp<"tile_eq"> {
```
Comment (vwellsTT) [inline include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td]: I think (but not 100% sure) you probably want to use TTIR_GenericRegionComputeBinaryDstOp instead

Comment (dlokeTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td]: Yeah, making changes to make it an SFPU op

File: lib/Dialect/TTIR/Transforms/InsertDstRegisterAccess.cpp
Comment (dlokeTT) [inline lib/Dialect/TTIR/Transforms/InsertDstRegisterAccess.cpp:1]: Changes to this file are in this pr #4845, related to writing intermediate results to dst register between tile ops in the same linalg generic. That one should land first but need the fixes here to pass all the tests for comparison ops.

File: test/python/golden/test_ttir_ops.py [hunk 2]
```diff
@@ -2334,6 +2334,74 @@ def test_unary_ops_int32(
...
+        elif test_fn.__name__ == "ne":
+            golden_output = torch.ne(input_tensor1, input_tensor2).to(dtype)
+        elif test_fn.__name__ == "le":
+            golden_output = torch.le(input_tensor1, input_tensor2).to(dtype)
+        elif test_fn.__name__ == "lt":
+            golden_output = torch.lt(input_tensor1, input_tensor2).to(dtype)
+        elif test_fn.__name__ == "ge":
+            golden_output = torch.ge(input_tensor1, input_tensor2).to(dtype)
+        elif test_fn.__name__ == "gt":
+            golden_output = torch.gt(input_tensor1, input_tensor2).to(dtype)
```
Comment (tapspatel) [inline test/python/golden/test_ttir_ops.py]: is there a reason we are calling these golden functions instead of using builder golden functions?

File: test/python/golden/test_ttir_ops.py [hunk 3]
```diff
@@ -2334,6 +2334,74 @@ def test_unary_ops_int32(
     )
 
 
+@pytest.mark.parametrize("shape", [(128, 128)], ids=shape_str)
+@pytest.mark.parametrize("dtype", [torch.float32], ids=["f32"])
+@pytest.mark.parametrize("target", ["ttnn", "ttmetal"])
+@pytest.mark.parametrize(
+    "test_fn",
```
Comment (tapspatel) [inline test/python/golden/test_ttir_ops.py]: any reason why this was moved outside of `test_unary_ops_int32`

Comment (dlokeTT) [reply test/python/golden/test_ttir_ops.py]: I took them out of test_binary_ops because I wanted to set the golden inputs and outputs to test edge cases and same reason for why I didn't call the builder golden functions because I need the shape for creating the inputs

**Comment by nsmithtt** [comment]
FYI reviewers this PR depends on:
- https://github.com/tenstorrent/tt-mlir/pull/4845

File: test/python/golden/test_ttir_ops.py (hunk 2 above)
Comment (tapspatel) [reply test/python/golden/test_ttir_ops.py]: just in time: https://github.com/tenstorrent/tt-mlir/commit/229ee450f056090c31f63a944394006a8f4bc0ed

You don't need to explicity set the golden output. You can just set your inputs and builder goldens will now use those in golden calculation

File: test/python/golden/test_ttir_ops.py (hunk 3 above)
Comment (tapspatel) [reply test/python/golden/test_ttir_ops.py]: gotcha

File: test/python/golden/test_ttir_ops.py (hunk 2 above)
Comment (dlokeTT) [reply test/python/golden/test_ttir_ops.py]: sweet, will use that now!

Comment (dlokeTT) [reply test/python/golden/test_ttir_ops.py]: updated

**Review by tapspatel — APPROVED** [review]
builder and tests look good

File: include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td [hunk 4]
```diff
@@ -280,6 +280,77 @@ def TTIR_TileLogicalNotOp : TTIR_GenericRegionComputeUnaryDstOp<"tile_logical_no
     let results = (outs TTCore_Tile:$result);
 }
 
+def TTIR_TileSubBinaryOp : TTIR_GenericRegionComputeBinaryDstOp<"tile_sub_binary"> {
```
Comment (vwellsTT) [inline include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td:283]: not sure I follow why this is different than a regular TileSub op?
(If a separate op is needed, I think we definitely could use a comment here explaining too)

Comment (dlokeTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td:283]: This is an sfpu op

Comment (dlokeTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td:283]: I can add a comment

Comment (vwellsTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td:283]: Ah, makes sense.  I don't love the name but I guess it's just copying the name from Metal, so prob best to leave as-is

File: lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp [hunk 5]
```diff
@@ -329,8 +329,15 @@ using ComputeOpMap = OpMap<
   std::pair<ttir::TileSigmoidOp,     std::pair<ttkernel::SigmoidTileInitOp,         ttkernel::SigmoidTileOp>>,
   std::pair<ttir::TileSinOp,         std::pair<ttkernel::SinTileInitOp,             ttkernel::SinTileOp>>,
   std::pair<ttir::TileTanOp,         std::pair<ttkernel::TanTileInitOp,             ttkernel::TanTileOp>>,
+  std::pair<ttir::TileEqzOp,         std::pair<ttkernel::EqzTileInitOp,             ttkernel::EqzTileOp>>,
```
Comment (vwellsTT) [inline lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp:332]: I'm a bit confused why the I32 variants aren't listed here too--they're just included for future-proofing, but not generate-able today? Or you expect to generate them post TTIR somehow?

File: lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp [hunk 6]
```diff
@@ -513,7 +520,9 @@ class TTIRSFPUOpsRewriter : public OpConversionPattern<ConcreteOp> {
     auto inCB = getInCB(rewriter, op);
     auto outCB = getOutCB(rewriter, op);
     setInsertionPointAfterOperands(rewriter, {inCB, outCB});
-    rewriter.create<ttkernel::InitSFPUOp>(op->getLoc(), inCB, outCB);
```
Comment (vwellsTT) [inline lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp:516]: why was this done unconditionally before, was that erroneous?

File: lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp [hunk 7]
```diff
@@ -556,6 +565,43 @@ class TTIRSFPUOpsRewriter : public OpConversionPattern<ConcreteOp> {
...
+      const auto elemType =
+          mlir::cast<ttcore::TileType>(op.getInput().getType())
+              .getElementType();
+      bool isCBI32 = false;
+      if (llvm::isa<IntegerType>(elemType)) {
+        isCBI32 = mlir::cast<IntegerType>(elemType).isSigned() &&
+                  mlir::cast<IntegerType>(elemType).getWidth() == 32;
+      }
+      if (isCBI32) {
+        if (std::is_same_v<SFPUOp, ttkernel::EqzTileOp>) {
```
Comment (vwellsTT) [inline lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp]: hmm, I wonder if there's a better way to write this.  I think the pattern
```
if(A || B || C)
{
  ...
  if (A) 
  {
    ...
  } else if (B)
  {
    ...
  } else if (C)
  ...
```
is a bit ugly, though I'm not sure if there's actually a cleaner way to write it here...

File: lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp (hunk 5 above)
Comment (dlokeTT) [reply lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp:332]: The type is deduced inside the sfpu ops rewriter. I followed how it was done before, this could probably get refactored to be cleaner since it's a bunch of if statements. Just left that for now

File: lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp (hunk 6 above)
Comment (dlokeTT) [reply lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp:516]: I think this was unconditional because we had never tried to have two ops in the same compute kernel. With comparisons now we do a sub_binary followed by the comparison. If we lower both of them without this, we get 2 init_sfpu calls in the final kernel. It doesn't break anything but is redundant and for this pr we only use the sub_binary along with a comparison op so I added that check.

File: lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp [hunk 8]
```diff
@@ -341,6 +341,126 @@ class TTIRNamedElementwiseRewriter final
...
+        // Create 'linalg.generic' accepting 'blockArgs'.
+
+        SmallVector<mlir::AffineMap> linalgIndexingMaps =
+            getAffineMapsArray(rewriter, numOperands, rank);
+        SmallVector<mlir::utils::IteratorType> linalgIteratorTypes =
+            iteratorTypeTTIRToLinalg(rewriter, iteratorTypes);
+
+        auto linalgGeneric = rewriter.create<mlir::linalg::GenericOp>(
+            loc,
+            /* result tensor types */
```
Comment (vwellsTT) [inline lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp]: nit: I don't like these comments styles, I think such comments should be like 
```
/*param_name=*/

File: lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp [hunk 9]
```diff
@@ -341,6 +341,126 @@ class TTIRNamedElementwiseRewriter final
...
+  matchAndRewrite(ConcreteOp op, typename ConcreteOp::Adaptor adaptor,
+                  mlir::ConversionPatternRewriter &rewriter) const final {
+    mlir::MLIRContext *ctx = rewriter.getContext();
+    mlir::Location loc = op->getLoc();
+
+    auto [origInputs, origOutputs] =
+        splitDpsSignature(adaptor, op.getDpsInits().size());
+    auto [inputs, outputs] =
+        toLayoutOperandsAndResults(rewriter, {origInputs, origOutputs},
+                                   /*tiled*/ true);
```
Comment (vwellsTT) [inline lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp]: nit: again subjective but personally pref `/*tiled=*/`

File: lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp (hunk 6 above)
Comment (dlokeTT) [reply lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp:516]: Curious to know what others think about this. If we end up keeping the check, I'll add a comment about it

Comment (vwellsTT) [reply lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp:516]: I don't have any problem with this, just wasn't sure on the context--that makes sense, cool

File: lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp (hunk 7 above)
Comment (vwellsTT) [reply lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp]: can't think of a better pattern for this case cuz of the inner `if(isCBI32)` check, so maybe this is fine

File: test/python/golden/test_ttir_ops.py [hunk 10]
```diff
@@ -2440,6 +2434,70 @@ def test_bitwise_binary_ops(test_fn: Callable, shape: Shape, request):
...
+    [
+        eq,
+        ne,
+        le,
+        lt,
+        ge,
+        gt,
+    ],
+)
+def test_binary_comparison_ops(
```
Comment (vwellsTT) [inline test/python/golden/test_ttir_ops.py]: nit: are there any non-binary comparison ops? If not personally would just call this `test_comparison_ops` I think (can't think of what a unary comparison op would be, but could be wrong)

**Review by vwellsTT — APPROVED** [review]
This looks solid to me, some nits + questions aside

File: lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp (hunk 7 above)
Comment (dlokeTT) [reply lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp]: 👍 , I have some ideas for clean up. Might attempt this in a future pr

File: include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td (hunk 4 above)
Comment (wenbinlyuTT) [reply include/ttmlir/Dialect/TTIR/IR/TTIRGenericRegionOps.td:283]: @dlokeTT Instead of the comment above, we can just specify it's SFPU-only in the op's description field.

File: include/ttmlir/Dialect/TTKernel/IR/TTKernelOps.td [hunk 11]
```diff
@@ -1128,6 +1128,279 @@ def TTKernel_LogicalNotUnaryTileI32Op : TTKernel_SFPUOp<"logical_not_unary_tile_
...
+
+    let arguments = (ins IndexLike:$dst0_index);
+
+    let assemblyFormat = [{
+      `(` $dst0_index `)` attr-dict `:` functional-type(operands, results)
+    }];
+}
+
+
+
```
Comment (wenbinlyuTT) [inline include/ttmlir/Dialect/TTKernel/IR/TTKernelOps.td]: Extra new lines.

File: lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp [hunk 12]
```diff
@@ -341,6 +341,126 @@ class TTIRNamedElementwiseRewriter final
...
 } // namespace
 
+namespace {
+// ----------------------------------------------------------------------------
+//
+// Rewrite comparison elementwise ops by emitting subtract binary tile op
+// followed by zero comparsion tile version of the op into a
+// ttir.generic/linalg.generic nest.
+template <typename ConcreteOp, typename TileOp>
+class TTIRNamedComparisonRewriter final
```
Comment (wenbinlyuTT) [inline lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp]: This rewriter is 95% identical to the existing ElementwiseRewriter.

I'm kind of torn here, keeping two rewriters increases maintenance burden, but unifying them will just add more ugly special cases.

Maybe we really need to sit down and think of some systematic way to model these "tiny" diverging code paths due to op/dataformat requirements.

File: lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp [hunk 13]
```diff
@@ -556,6 +568,43 @@ class TTIRSFPUOpsRewriter : public OpConversionPattern<ConcreteOp> {
...
+        } else if (std::is_same_v<SFPUOp, ttkernel::GtzTileOp>) {
+          rewriter.create<ttkernel::GtzTileI32Op>(op->getLoc(),
+                                                  adaptor.getInput());
+        } else if (std::is_same_v<SFPUOp, ttkernel::GezTileOp>) {
+          rewriter.create<ttkernel::GezTileI32Op>(op->getLoc(),
+                                                  adaptor.getInput());
+        } else if (std::is_same_v<SFPUOp, ttkernel::LtzTileOp>) {
+          rewriter.create<ttkernel::LtzTileI32Op>(op->getLoc(),
+                                                  adaptor.getInput());
+        } else if (std::is_same_v<SFPUOp, ttkernel::LezTileOp>) {
```
Comment (wenbinlyuTT) [inline lib/Conversion/TTIRToTTKernel/TTIRToTTKernel.cpp]: I think the if's in front of `std::is_same_v`'s can all be `constexpr`.

File: test/python/golden/test_ttir_ops.py [hunk 14]
```diff
@@ -2440,6 +2434,70 @@ def test_bitwise_binary_ops(test_fn: Callable, shape: Shape, request):
...
+        in1: Operand,
+        builder: TTIRBuilder,
+        unit_attrs: Optional[List[str]] = None,
+    ):
+        # Generate randn tensors and convert to float for input tensors
+        randn_tensor1 = torch.randn(shape, dtype=torch.float32)
+        randn_tensor2 = torch.randn(shape, dtype=torch.float32)
+
+        # Set some indices in randn_tensor2 to be the same as randn_tensor1
+        # This ensures we have both equal and unequal values for comprehensive testing
```
Comment (wenbinlyuTT) [inline test/python/golden/test_ttir_ops.py:2471]: IMHO, most of the comments in this test can be removed, sans the two lines above. The code is already pretty self-explanatory. As long as the intention of overwriting the golden inputs is made clear, there's no need to comment the other stuff.

File: lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp [hunk 15]
```diff
@@ -248,6 +248,15 @@ class TTIRNamedElementwiseRewriter final
                                 targetGridShape) {}
 
 private:
+  static bool isComparisonOp() {
```
Comment (wenbinlyuTT) [inline lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp]: If you do `static constexpr bool isComparisonOp = ...` here

Comment (nsmithtt) [inline lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp]: static constexpr bool

File: lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp [hunk 16]
```diff
@@ -309,9 +318,21 @@ class TTIRNamedElementwiseRewriter final
             linalgIteratorTypes,
             [&](mlir::OpBuilder &bbBuilder, mlir::Location bbLoc,
                 mlir::ValueRange bbArgs) {
-              mlir::Value yield = bbBuilder.create<TileOp>(
-                  loc, /* resultTypes */ bbArgs.take_back(numOutputs),
-                  /* operands */ bbArgs.take_front(numInputs));
+              mlir::Value yield;
+              if (isComparisonOp()) {
```
Comment (wenbinlyuTT) [inline lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp]: ... this can then become `if constexpr (isComparisonOp) {...`

File: lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp [hunk 17]
```diff
@@ -309,9 +318,21 @@ class TTIRNamedElementwiseRewriter final
...
+                    /*operands=*/bbArgs.take_front(numInputs));
+                yield = bbBuilder.create<TileOp>(
+                    loc, /* resultTypes */ bbArgs.take_back(numOutputs),
+                    /* operands */ subResult);
+              } else {
+                // For regular elementwise ops, create TileOp directly
+                yield = bbBuilder.create<TileOp>(
+                    loc, /* resultTypes */ bbArgs.take_back(numOutputs),
+                    /* operands */ bbArgs.take_front(numInputs));
+              }
```
Comment (nsmithtt) [inline lib/Conversion/TTIRToTTIRGeneric/TTIRToTTIRGeneric.cpp:335]: Could be written more (_slightly_) succinctly:
```c++
              mlir::ValueRange operands = bbArgs.take_front(numInputs);
              if (isComparisonOp()) {
                // For comparison ops, first subtract then compare with zero
                operands = bbBuilder.create<ttir::TileSubBinaryOp>(
                    loc, /*resultTypes=*/bbArgs.take_back(numOutputs), operands);
              }

              mlir::Value yield = bbBuilder.create<TileOp>(
                  loc, /* resultTypes */ bbArgs.take_back(numOutputs), operands);
```

**Review by nsmithtt — APPROVED** [review]
LGTM! Optional improvements inline:
